src/core/truthTable/truth_table.cpp

```diff
@@ -29,38 +29,45 @@ namespace syrec {
         return result;
     }
 
-    auto TruthTable::equal(TruthTable& tt1, TruthTable& tt2, bool equalityUpToDontCare) -> bool {
+    auto TruthTable::equal(TruthTable const& tt1, TruthTable const& tt2, bool equalityUpToDontCare) -> bool {
         if (!equalityUpToDontCare) {
             return (tt1 == tt2);
         }
 
-        if (tt1.nInputs() != tt2.nInputs()) {
+        // both the truth table must have the same number of I/O combinations.
+        if (tt1.size() != tt2.size()) {
             return false;
         }
 
-        const auto nBits = tt1.nInputs();
-        assert(nBits < 65U);
-        const auto totalInputs = 1U << nBits;
+        // no. of primary outputs.
+        const auto noOfPrimaryOutputs = static_cast<int>(std::min(tt1.nOutputs(), tt2.nOutputs()));
 
-        std::uint64_t n = 0U;
+        auto tt1It = tt1.begin();
+        auto tt2It = tt2.begin();
 
-        while (n < totalInputs) {
-            const auto input = TruthTable::Cube::fromInteger(n, nBits);
-            ++n;
+        while (tt1It != tt1.end() || tt2It != tt2.end()) {
+            TruthTable::Cube const out1{tt1It->second.begin(), tt1It->second.begin() + noOfPrimaryOutputs};
+            TruthTable::Cube const out2{tt2It->second.begin(), tt2It->second.begin() + noOfPrimaryOutputs};
```

Comment:
it is not guaranteed that the primary inputs are the first or the last entries in the truth table. For each cube from a TT the proper "primary input pattern" and the "primary output pattern" need to be determined based on the constants and garbage of the respective TT. Sounds like a perfect use-case for a member function in the TT class, that returns the "filtered" input / output cube.